File: backend/app/schemas/text_library.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.text_library import (
    TextImportJobStatus,
    TextImportTaskStatus,
    TextItemFieldStatus,
    TextSourceChannel,
)


class TextLibraryItemResponse(BaseModel):
    id: int
    name: str
    content: str
    source_channel: TextSourceChannel
    title_status: TextItemFieldStatus
    content_status: TextItemFieldStatus
    processing_stage: str | None = None
    processing_message: str | None = None
    error_message: str | None = None
    source_url: str | None = None
    source_file_name: str | None = None
    source_post_id: str | None = None
    source_post_updated_at: str | None = None
    is_enabled: bool
    created_at: datetime
    updated_at: datetime


class TextLibraryListResponse(BaseModel):
    items: list[TextLibraryItemResponse]
    total: int
    page: int | None = None
    page_size: int | None = None


class TextLibraryUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    content: str | None = Field(default=None, min_length=1)
    is_enabled: bool | None = None


class TextLibraryImportCopyRequest(BaseModel):
    content: str = Field(..., min_length=1)


class TextLibraryImportLinksRequest(BaseModel):
    urls_text: str = Field(..., min_length=1)


class TextLibraryImportLinksCreateResponse(BaseModel):
    job_id: str
    job_ids: list[str] = Field(default_factory=list)
    item_ids: list[int] = Field(default_factory=list)


class TextLibraryImportTaskResponse(BaseModel):
    id: int
    source_url: str
    source_channel: TextSourceChannel
    status: TextImportTaskStatus
    stage: str | None = None
    stage_message: str | None = None
    cache_hit: bool = False
    error_message: str | None = None
    cancel_requested: bool = False
    cancel_requested_at: datetime | None = None
    cancel_reason: str | None = None
    retry_of_task_id: int | None = None
    retry_no: int = 0
    text_library_item_id: int | None = None
    source_post_id: str | None = None
    source_post_updated_at: str | None = None


class TextLibraryImportJobResponse(BaseModel):
    id: str
    status: TextImportJobStatus
    total_count: int
    completed_count: int
    success_count: int
    failed_count: int
    canceled_count: int = 0
    cancel_requested: bool = False
    cancel_requested_at: datetime | None = None
    cancel_requested_by: str | None = None
    terminal_at: datetime | None = None
    error_message: str | None = None
    tasks: list[TextLibraryImportTaskResponse]
    created_at: datetime
    updated_at: datetime


class TextLibraryCancelResponse(BaseModel):
    affected_jobs: int = 0
    affected_tasks: int = 0

File: frontend/src/types/text-library.ts
export type TextSourceChannel = 'copy' | 'file' | 'web' | 'xiaohongshu' | 'douyin' | 'kuaishou'
export type TextImportJobStatus = 'pending' | 'running' | 'completed' | 'failed' | 'canceled'
export type TextImportTaskStatus = 'pending' | 'running' | 'completed' | 'failed' | 'canceled'
export type TextItemFieldStatus = 'pending' | 'running' | 'ready' | 'failed' | 'canceled'

export interface TextLibraryItem {
  id: number
  name: string
  content: string
  source_channel: TextSourceChannel
  title_status: TextItemFieldStatus
  content_status: TextItemFieldStatus
  processing_stage?: string | null
  processing_message?: string | null
  error_message?: string | null
  source_url?: string | null
  source_file_name?: string | null
  source_post_id?: string | null
  source_post_updated_at?: string | null
  is_enabled: boolean
  created_at: string
  updated_at: string
}

export interface TextLibraryListResponse {
  items: TextLibraryItem[]
  total: number
  page?: number | null
  page_size?: number | null
}

export interface TextLibraryImportTask {
  id: number
  source_url: string
  source_channel: TextSourceChannel
  status: TextImportTaskStatus
  stage?: string | null
  stage_message?: string | null
  cache_hit?: boolean
  error_message?: string | null
  cancel_requested?: boolean
  cancel_requested_at?: string | null
  cancel_reason?: string | null
  retry_of_task_id?: number | null
  retry_no?: number
  text_library_item_id?: number | null
  source_post_id?: string | null
  source_post_updated_at?: string | null
}

export interface TextLibraryImportJob {
  id: string
  status: TextImportJobStatus
  total_count: number
  completed_count: number
  success_count: number
  failed_count: number
  canceled_count?: number
  cancel_requested?: boolean
  cancel_requested_at?: string | null
  cancel_requested_by?: string | null
  terminal_at?: string | null
  error_message?: string | null
  tasks: TextLibraryImportTask[]
  created_at: string
  updated_at: string
}

export interface TextLibraryImportLinksCreateResponse {
  job_id: string
  job_ids?: string[]
  item_ids: number[]
}

export interface SourceImportFromTextLibraryResultItem {
  text_library_id: number
  status: 'created' | 'skipped' | 'failed'
  source_id?: number | null
  message: string
}

export interface SourceImportFromTextLibraryResponse {
  success: boolean
  summary: {
    requested_count: number
    created_count: number
    skipped_count: number
    failed_count: number
  }
  results: SourceImportFromTextLibraryResultItem[]
}

export interface TextLibraryCancelResponse {
  affected_jobs: number
  affected_tasks: number
}
